### PR TITLE
Fixed gasp support bugs per #738

### PR DESCRIPTION
**Pull request by Balearica**
## Context
See #738.  In short, write support for the `gasp` table was implemented recently in #595, however is broken.  Any font written by opentype.js that includes a `gasp` table cannot be re-imported by opentype.js.  

## Changes
This PR makes two changes to resolve.  First, it fixes the typo in the `gasp` write function, which was trying to iterate over the integer `gasp.numRanges` rather than the array `gasp.gaspRanges`.  Second, it wraps the entire `gasp` read function in a `try` block.  This should prevent any future issues where failing to parse this optional table prevents a font from being read.  

## How Has This Been Tested?
I confirmed that the [Changa-Regular.ttf](https://github.com/opentypejs/opentype.js/blob/master/test/fonts/Changa-Regular.ttf) font could be saved/read using `toArrayBuffer`/`opentype.parse`, which (as described in #738) previously resulted in an error.  

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I did `npm run test` and all tests passed green (including code styling checks).
- [ ] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the **README** accordingly.
- [x] I have read the **Contribute** README section.

### CODE REVIEW

**Comment by Connum** [comment]
Please add a test that will prevent this issue from being reintroduced in the future.

**Comment by Balearica** [comment]
I added a test that confirms fonts written with a `gasp` table can be read, and that the `gasp` table is identical to the original.

**Comment by Connum** [comment]
> I added a test that confirms fonts written with a `gasp` table can be read, and that the `gasp` table is identical to the original.

Very nice, thanks! We shall have a review on this fix soon.

File: test/tables/gasp.spec.mjs
Comment (yne) [inline test/tables/gasp.spec.mjs:34]: we are using JSON because there are no assert functions that can do a deep compare ?

File: src/opentype.mjs
Comment (yne) [inline src/opentype.mjs:313]: not sure if the try/catch is still needed here
and if it is needed, maybe we shall have it for every cases to keep a unified behavior

**Review by yne — APPROVED** [review]
thanks for the fix :)
just 2 simple questions/suggestion but no issue overall

File: src/opentype.mjs
Comment (Balearica) [reply src/opentype.mjs:313]: I believe that increasing the scope of OpenType.js to include a bunch of optional tables necessitates making them fail gracefully if they cannot be parsed.  The primary issue in my case is not that `gasp` could not be parsed correctly (I don't care about this table at all), but rather that my program crashed simply because this optional table could not be parsed.  Even after the fixes in this PR, exceptions during parsing this table may occur due to (1) undiscovered bugs in the parsing/writing functions or (2) flaws in the `gasp` table of otherwise valid user-provided fonts.  FontForge handles failures when parsing optional tables by skipping and printing a warning message, which I think makes sense. 

If you want, I can update such that all optional tables are handled in a similar way.  I only added for `gasp` as I see that table as higher-risk (as it was added the most recently without being fully tested), but I don't think crashing when an optional table cannot be parsed is ever a desirable behavior.

File: test/tables/gasp.spec.mjs
Comment (Balearica) [reply test/tables/gasp.spec.mjs]: I modified to use `assert.deepStrictEqual`.

File: src/opentype.mjs
Comment (yne) [reply src/opentype.mjs:313]: I would have expected the inner parse function to do the console.war/error + return null
But I just saw that `gasp.parse` had a `check.argument` that could `throw`
so my comment is irrelevant and the try/catch is well used here.
And I agree with you that we shall be error-resilient.

Comment (Connum) [reply src/opentype.mjs:313]: > I believe that increasing the scope of OpenType.js to include a bunch of optional tables necessitates making them fail gracefully if they cannot be parsed.

Graceful table handling will be implemented via #704

**Comment by Connum** [comment]
Hi, (why) are we waiting for another review? It looks to me as if this is ready to be merged.

**Comment by yne** [comment]
> Hi, (why) are we waiting for another review? It looks to me as if this is ready to be merged.

I've approved but the repo owner seems to want 2 approval, and @ILOVEPIE don't seems available to approve.

can't you bypass the process and force the merge ?

**Pull request merged**
